File: tests/lua/write.lua
-- please install https://github.com/mah0x211/lua-process
-- this is necessary to use getpid();

process = require('process')

-- http://lua-users.org/wiki/SimpleRound
function round(num, idp)
  local mult = 10^(idp or 0)
  return math.floor(num * mult + 0.5) / mult
end

-- This function used to be a generic function to call any system command
-- but was bugging and now only capture memory usage

function capture_memory_usage(pid)

  local command = ("ps -o rss= -p :pid"):gsub(':pid', pid)

  local io = io.popen(command, 'r')
  local f = assert(io)
  local s = assert(f:read('*a'))
  f:close()
  if raw then return s end
  s = string.gsub(s, '^%s+', '')
  s = string.gsub(s, '%s+$', '')
  s = string.gsub(s, '[\n\r]+', ' ')

  return s
end

function print_memory_usage(memory_before, memory_after)
  local memory_usage = round(((memory_after - memory_before) / 1024.0), 2)

  local memory_str = ("Memory: :memory_usage MB"):gsub(':memory_usage', memory_usage)
  print(memory_str)
end

function print_time_spend(time)
  -- os.clock():
  -- Returns an approximation of the amount in seconds of CPU time used by the program.
  local time_str = ("Time: :timems"):gsub(':time', time * 1000)
  print(time_str)
end

function test_type_label(test_type, test_name, test_mode)
  local str=[=[
Test Language: Lua
Test Language Version: :version
Test Type: :test_type
Test Name: :test_name
Test Mode: :test_mode]=]
  str = str:gsub(':version', io.popen('lua -v'):read())
  str = str:gsub(':test_type', test_type)
  str = str:gsub(':test_name', test_name)
  str = str:gsub(':test_mode', test_mode)

  return str
end

BYTES_PER_LINE = 1024
DUMMY_CHARACTER = " "

function generate_dummy_data(block_size)
  str = string.rep(DUMMY_CHARACTER, (block_size * BYTES_PER_LINE))

  return (str .. "\n")
end

function test_stream(file_path, block_size)
  local file = io.open(file_path, "w")

  for i=0,block_size,1 do
    file:write(generate_dummy_data(1))
  end

  file:flush()
  file:close()
end

function test_full(file_path, block_size)
  local file = io.open(file_path, "w")

  file:write(generate_dummy_data(block_size))
  file:flush()
  file:close()
end

function basename(str)
  value = str:gsub("(.*/)(.*)", '%2')

  return value
end

-- get options from system env(MODE=stream VERBOSE=true lua lua/read.lua)
local file_path = os.getenv('TEST_FILE')
local test_type = os.getenv('TEST_TYPE')
local test_mode = (os.getenv('TEST_MODE') or 'full')
local block_size = tonumber(os.getenv('KB_BLOCK_SIZE') or 1)

local test_name = (os.getenv('TEST_NAME') or basename(file_path))
local verbose = (os.getenv('VERBOSE') == 'true')

function run(file_path, test_mode, test_type, test_name, verbose)
  if verbose then
    print(test_type_label(test_type, test_name, test_mode))
  end

  local memory_before = capture_memory_usage(process.getpid())

  if(test_mode == 'full') then
    test_full(file_path, block_size)
  else
    test_stream(file_path, block_size)
  end

  local memory_after = capture_memory_usage(process.getpid())

  if verbose then
    print_memory_usage(memory_before, memory_after)
    print_time_spend(os.clock())
  end
end

run(file_path, test_mode, test_type, test_name, verbose)

File: tests/go/write.go
package main

import (
 "os"
 "fmt"
 "time"
 "math"
 "bufio"
 "runtime"
 "strings"
 "strconv"
// "log"
// "io/ioutil"
)

func check(e error) {
    if e != nil {
        panic(e)
    }
}

func generate_dummy_data(block_size int) []byte {
  DUMMY_CHARACTER := " "
  BYTES_PER_LINE := 1024

  str := strings.Repeat(DUMMY_CHARACTER, (BYTES_PER_LINE * block_size))

  return []byte(str)
}


func test_full(file_path string, block_size int) {
  content := generate_dummy_data(block_size)
  file, err := os.OpenFile(file_path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)

  check(err)

  _, _err := file.Write(content)

  check(_err)

  file.Close()
}

func test_stream(file_path string, block_size int) {
  file, err := os.OpenFile(file_path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)

  check(err)

  w := bufio.NewWriter(file)

  for i := 0; i < block_size; i++ {
    content := generate_dummy_data(1)
    w.WriteString(string(content))
  }

  w.Flush()
  file.Close()
}

func Round(f float64) int {
    if math.Abs(f) < 0.5 {
        return 0
    }
    return int(f + math.Copysign(0.5, f))
}

func print_memory_usage(memory_before, memory_after uint64) {
  memory := ((memory_after - memory_before) / 1024.0)
  memory_usage := Round(float64(memory))

  fmt.Println(fmt.Sprintf("Memory: %d MB", memory_usage))
}

/* This basically parse /proc/PID/status to get RAM usage */
/* In others languages benchmarks, this is achieved by running: ps -o rss= -p PID */
func capture_memory_usage() uint64 {
  var mem runtime.MemStats
  runtime.ReadMemStats(&mem)

  return (mem.Alloc / 1024)
}

func print_time_spend(time string) {
  fmt.Println(fmt.Sprintf("Time: %s", time))
}

func print_test_type(test_type string, test_mode string, test_name string) {
  str := `
Test Language: GoLang
Test Language Version: go version go1.7.3 linux/amd64
Test Type: %s
Test Name: %s
Test Mode: %s
`

  fmt.Print(fmt.Sprintf(str, test_type, test_name, test_mode))
}

func main() {
  file := os.Getenv("TEST_FILE")
  verbose := os.Getenv("VERBOSE") == "true"
  test_mode := os.Getenv("TEST_MODE")
  test_name := os.Getenv("TEST_NAME")
  test_type := os.Getenv("TEST_TYPE")
  block_size, _ := strconv.Atoi(os.Getenv("KB_BLOCK_SIZE"))

  if test_mode == "" { test_mode = "full" }

  if test_type == "" {
    panic("You must suply TEST_TYPE with one the values: read, write, read+write")
  }

  start := time.Now()

  memory_before := capture_memory_usage()

  if verbose == true {
    print_test_type(test_type, test_mode, test_name)
  }

  if test_mode == "full" {
    test_full(file, block_size)
  } else if test_mode == "stream" {
    test_stream(file, block_size)
  }

  memory_after := capture_memory_usage()

  elapsed := time.Since(start)

  if verbose == true {
    print_time_spend(elapsed.String());
    print_memory_usage(memory_before, memory_after)
  }
}